### PR TITLE
Add basic .gitlab-ci.yml file

### PR DESCRIPTION
**Pull request by msgilligan**
Add a GitLab CI build using Ubuntu Bionic and OpenJDK 11

### CODE REVIEW

**Comment by schildbach** [comment]
What's the purpose of this? As long as PRs are opened on GitHub (and not GitLab), this CI will not run.

**Comment by msgilligan** [comment]
If you mirror the repository to Gitlab it will run.

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+image: ubuntu:bionic
+
+variables:
+  GRADLE_VERSION: "5.4.1"
```
Comment (schildbach) [inline .gitlab-ci.yml]: I think for now we should stay with the same Gradle version as in .travis.yml: 4.10.3

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+before_script:
+  - apt-get update && apt-get -y upgrade
+  - apt-get -y install openjdk-11-jdk gradle
+  - gradle wrapper --gradle-version=$GRADLE_VERSION
+  - chmod +x ./gradlew
+
+build:
+  script:
+    - java -version
+    - LC_ALL=C.UTF-8 ./gradlew build
```
Comment (schildbach) [inline .gitlab-ci.yml]: Is the environment variable necessary? If yes, why? If not, can we remove the assignment?

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  artifacts:
+    paths:
+      - core/build
```
Comment (schildbach) [inline .gitlab-ci.yml]: What does Gitlab do with these artifacts?

File: .gitlab-ci.yml (hunk 1 above)
Comment (msgilligan) [reply .gitlab-ci.yml]: Sure, I'll change it. (At some point we might want to do builds with multiple Gradle versions as we are doing on Travis)

File: .gitlab-ci.yml (hunk 2 above)
Comment (msgilligan) [reply .gitlab-ci.yml]: It's needed to prevent fatal errors generating JavaDoc due to character set. There may be another, better way of resolving the issue, but I'm pretty sure if I remove this it will be broken. I'll verify.

File: .gitlab-ci.yml (hunk 3 above)
Comment (msgilligan) [reply .gitlab-ci.yml]: Makes the available for download. (On commercial projects with private gitlab, I've used the CI-built artifacts as the official build binaries)

File: .gitlab-ci.yml (hunk 2 above)
Comment (schildbach) [reply .gitlab-ci.yml]: Hmm, why is this not an issue locally or on Travis?

Maybe adding

```
javadoc {
    options.encoding = 'UTF-8'
}
```

to build.gradle could also help?

File: .gitlab-ci.yml (hunk 3 above)
Comment (schildbach) [reply .gitlab-ci.yml]: Ok. I'm not sure if we want binary downloads on GitLib (or GitHub) however. Our website downloads currently point at maven.org.

File: .gitlab-ci.yml (hunk 1 above)
Comment (schildbach) [reply .gitlab-ci.yml]: Yes, is there something like the `matrix` availble?

**Comment by msgilligan** [comment]
One cool thing about GitLab CI is that you can run it locally with a command like:
```
 gitlab-runner exec docker --docker-privileged build
```

**Comment by msgilligan** [comment]
@schildbach I started using GitLab on my wallet-framework branch because I wanted to run CI with Java 12 (although I've since dialed it back to 11.) You asked me to provide GitLab support as a PR against `master` -- did you have anything in particular in mind?

File: .gitlab-ci.yml (hunk 3 above)
Comment (msgilligan) [reply .gitlab-ci.yml]: Yeah, we can turn them off if we want. I also sometimes configure it so that other output files such as logs or debug/error information is preserved. They're not super-easy to find on the GitLab site, so I'm not too worried that people will get confused by them and use them instead of the official Maven binaries.

File: .gitlab-ci.yml (hunk 1 above)
Comment (msgilligan) [reply .gitlab-ci.yml]: There is, so I can try to add that in a subsequent PR. I just wanted to get a basic build working first.

File: .gitlab-ci.yml (hunk 2 above)
Comment (msgilligan) [reply .gitlab-ci.yml]: Trying that now...

**Comment by msgilligan** [comment]
> If you mirror the repository to Gitlab it will run.

To clarify: Not for PRs, but for any branches that are part of the mirror.

File: .gitlab-ci.yml (hunk 2 above)
Comment (schildbach) [reply .gitlab-ci.yml]: Did it work?

File: .gitlab-ci.yml (hunk 3 above)
Comment (schildbach) [reply .gitlab-ci.yml]: For a first step, I think I'd like to keep things simple and omit this part.

**Comment by schildbach** [comment]
> You asked me to provide GitLab support as a PR against `master` -- did you have anything in particular in mind?

Nothing particular, other than I thought running the unit tests on two different CIs is desirable. And maybe in future we can use those two CIs to verify reproducible builds.

**Comment by msgilligan** [comment]
I agree with you that two different CIs is desirable and I have done that on other projects. I also like the flexibility of GitLab CI's integration with Docker.

File: .gitlab-ci.yml (hunk 2 above)
Comment (msgilligan) [reply .gitlab-ci.yml]: It didn't work. (and I got distracted by something else, sorry.)

Can we go ahead and merge this as is, and fix it in a separate PR when we find a better way?

File: .gitlab-ci.yml (hunk 3 above)
Comment (msgilligan) [reply .gitlab-ci.yml]: Fixed in force-push.

**Comment by msgilligan** [comment]
@schildbach If you're ok with the `LC_ALL=C.UTF-8` (at least for the time-being) then this PR is ready to merge.

File: empty-settings.gradle [hunk 4]
```diff
@@ -0,0 +1,2 @@
+// Empty settings file to ignore subprojects
```
Comment (schildbach) [inline empty-settings.gradle]: This file seems to be unrelated to the PR. Can you remove it for now?

File: .gitlab-ci.yml (hunk 2 above)
Comment (schildbach) [reply .gitlab-ci.yml]: Well, I'm not too happy with "shell foo" in build scripts but ok, for now I'm fine with it.

File: empty-settings.gradle (hunk 4 above)
Comment (msgilligan) [reply empty-settings.gradle]: Actually, see line 9 of `.gitlab-ci.yml ` -- it solves the problem of the built-in Gradle not being able to process all the subproject of our `build.gradle`.

File: .gitlab-ci.yml (hunk 2 above)
Comment (schildbach) [reply .gitlab-ci.yml]: Weird, I just reproduced your issue by removing the environment varable assignment. Then I fixed the issue by adding `javadoc.options.encoding = 'UTF-8'` to the core build.gradle, which should be identical to my above suggestion.

Comment (msgilligan) [reply .gitlab-ci.yml]: OK, I'll rebase and see what happens.

File: empty-settings.gradle (hunk 4 above)
Comment (msgilligan) [reply empty-settings.gradle]: @schildbach Do yo see a better way of solving this issue?

File: .gitlab-ci.yml [hunk 5]
```diff
@@ -0,0 +1,15 @@
+image: ubuntu:bionic
+
+variables:
+  GRADLE_VERSION: "4.10.3"
+
+before_script:
+  - apt-get update && apt-get -y upgrade
```
Comment (schildbach) [inline .gitlab-ci.yml]: This line could be split into two lines for easier reading.

File: .gitlab-ci.yml [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+  - apt-get -y install openjdk-11-jdk gradle
+  - gradle -c empty-settings.gradle wrapper --gradle-version=$GRADLE_VERSION --info --stacktrace
+  - chmod +x ./gradlew
```
Comment (schildbach) [inline .gitlab-ci.yml]: This line isn't necessary; `gradlew` runs without.

File: .gitlab-ci.yml [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+build:
+  script:
+    - java -version
+    - LC_ALL=C.UTF-8 ./gradlew build --info --stacktrace
```
Comment (schildbach) [inline .gitlab-ci.yml]: I think we should remove the `--info` from here and above; it creates very long output that GitLab can only hardly handle.

File: empty-settings.gradle (hunk 4 above)
Comment (schildbach) [reply empty-settings.gradle]: Somehow I don't see the problem. For example see https://gitlab.com/bitcoinj/bitcoinj/-/jobs/260248387, it has run all the subprojects despite me having removed the empty-settings.gradle. Or am I missing something?

File: .gitlab-ci.yml (hunk 2 above)
Comment (msgilligan) [reply .gitlab-ci.yml]: Yeah, `javadoc.options.encoding = 'UTF-8'` fixes the build failure (don't recall exactly why my previous attempt at setting javadoc options failed.

However, the test output still has character set issues, so maybe we need to add one more setting in `build.gradle`. Also, we probably want to remove/replace the `--info` parameter to the two Gradle commands.

File: empty-settings.gradle (hunk 4 above)
Comment (msgilligan) [reply empty-settings.gradle]: OK, well that's cool. I probably changed something else that made the issue go away. It's likely that I had some other changes to the build files that weren't compatible with the gradle installed with apt-get. We may run into that in the future. The problem is that running `gradle wrapper` on a project requires the current version of gradle to be able to successfully process the root `build.gradle` and all the subprojects.

I'll go head and remove it.

File: .gitlab-ci.yml (hunk 5 above)
Comment (msgilligan) [reply .gitlab-ci.yml]: OK.

File: .gitlab-ci.yml (hunk 6 above)
Comment (msgilligan) [reply .gitlab-ci.yml]: OK.

File: .gitlab-ci.yml (hunk 7 above)
Comment (msgilligan) [reply .gitlab-ci.yml]: Agree, once everything else is debugged we should definitely remove it.

File: .gitlab-ci.yml (hunk 6 above)
Comment (msgilligan) [reply .gitlab-ci.yml]: Removed.

File: .gitlab-ci.yml (hunk 7 above)
Comment (msgilligan) [reply .gitlab-ci.yml]: OK, everything seems to be working -- both locally with Docker/gitlab-runner on macOS and on hosted Gitlab. So I've removed `--info`.

File: .gitlab-ci.yml (hunk 2 above)
Comment (msgilligan) [reply .gitlab-ci.yml]: > However, the test output still has character set issues

I was seeing `???` in the logging timestamps from the tests when running with Docker and `gitlab-runner` under macOS, but this is no longer happening.

We still might want to figure out how to set UTF-8 as the default character set at the system or Gradle level rather than task by task, but we can worry about that later if we need to.

**Comment by msgilligan** [comment]
@schildbach OK, all issues resolved. Should be ready to merge.

**Comment by msgilligan** [comment]
Travis failure looks like one of our normal intermittent failures...

**Comment by schildbach** [comment]
Merged! Thanks a lot for your patience with me.

**Comment by msgilligan** [comment]
> Thanks a lot for your patience with me.

It's a much better PR because of your work and feedback!
